File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slowpt = fastpt = head
        while fastpt and fastpt.next: 
            slowpt = slowpt.next
            fastpt = fastpt.next.next
            if slowpt == fastpt:
                return True
        return False

File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_map<ListNode*, int>dict;
        while(head != NULL){
            if(dict[head] == 1){
                return true;
            }
            else{
                dict[head] = 1;
            }
            head = head->next;
        }
        return false;
    }
};